packages/react/src/modules/slack/hooks/useSlackChannels.ts

```diff
@@ -1,77 +1,87 @@
-  tenant: string,
-  connectionsObject: ContainerObject,
-  knockSlackChannelId: string,
-  shouldRefetch: boolean,
-  setErrorMessage: (errorMessage: string | null) => void,
-  errorMessage: string | null,
-): {
-  data: { channels: Channel[] };
+type SlackChannelQueryOptions = {
+  maxCount?: number;
+  limitPerPage?: number;
+  excludeArchived?: boolean;
+  types?: string;
+  teamId?: string;
+};
+
+type UseSlackChannelsProps = {
+  tenant: string;
+  connectionsObject: ContainerObject;
+  queryOptions?: SlackChannelQueryOptions;
+  knockSlackChannelId: string;
+};
+
+type UseSlackChannelOutput = {
+  data: SlackChannel[];
   isLoading: boolean;
-  error: boolean;
-} {
-  const [isLoading, setIsLoading] = useState(true);
-  const [data, setData] = useState<{ channels: Channel[] }>({
-    channels: [],
-  });
+  refetch: () => void;
+};
 
+export function useSlackChannels({
+  tenant,
+  connectionsObject,
+  knockSlackChannelId,
+  queryOptions,
+}: UseSlackChannelsProps): UseSlackChannelOutput {
   const knockClient = useKnockClient();
 
-  useEffect(() => {
-    if (!errorMessage) {
-      const fetchChannels = async () => {
-        try {
-          await knockClient.slack
-            .getChannels({
-              tenant,
-              connectionsObject,
-              knockChannelId: knockSlackChannelId,
-            })
-            .then((res) => {
-              if (res.code === "ERR_BAD_REQUEST") {
-                const message = res.response?.data?.message
-                setData({ channels: [] });
-                setErrorMessage(message);
-                setIsLoading(false);
-                return;
-              }
+  const fetchChannels = ({ pageParam }: { pageParam: string }) => {
+    return knockClient.slack.getChannels({
+      tenant,
+      connectionsObject,
+      knockChannelId: knockSlackChannelId,
+      queryOptions: {
+        ...queryOptions,
+        cursor: pageParam,
+        limit: queryOptions?.limitPerPage || LIMIT_PER_PAGE,
+      },
+    });
+  };
+
+  const { data, isLoading, isFetching, fetchNextPage, hasNextPage, refetch } =
+    useInfiniteQuery({
+      queryKey: ["slackChannels"],
+      queryFn: fetchChannels,
+      initialPageParam: "",
+      getNextPageParam: (lastPage) =>
+        lastPage?.next_cursor === "" ? null : lastPage?.next_cursor,
+    });
 
-              setData({ channels: res.channels });
-              setErrorMessage(null);
-              setIsLoading(false);
-              return;
-            });
-        } catch (error) {
-          setErrorMessage("Something went wrong.");
-        }
-      };
+  const slackChannels =
+    data?.pages
+      ?.flatMap((page) => page?.slack_channels)
+      .filter((channel) => !!channel) || [];
 
-      setIsLoading(true);
-      fetchChannels();
+  const maxCount = queryOptions?.maxCount || MAX_COUNT;
+
+  useEffect(() => {
+    if (hasNextPage && !isFetching && slackChannels?.length < maxCount) {
+      fetchNextPage();
```

Comment:
this makes sure to continue making the query until we reach the maximum amount of requested channels